spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/types/TimestampParser.scala

```diff
@@ -32,30 +33,59 @@ object TimestampParser {
     16 -> "yyyy-MM-dd HH:mm",
     19 -> "yyyy-MM-dd HH:mm:ss",
     21 -> "yyyy-MM-dd HH:mmZ",
-    24 -> "yyyy-MM-dd HH:mm:ssZ",
+    22 -> "yyyy-MM-dd HH:mmZ",
     23 -> "yyyy-MM-dd HH:mm:ss.SSS",
-    28 -> "yyyy-MM-dd HH:mm:ss.SSSZ"
-  ).mapValues(DateTimeFormat.forPattern)
-
-  private val formatsWithTBySize = HashMap[Int, String](
-    16 -> "yyyy-MM-dd'T'HH:mm",
-    19 -> "yyyy-MM-dd'T'HH:mm:ss",
-    21 -> "yyyy-MM-dd'T'HH:mmZ",
-    24 -> "yyyy-MM-dd'T'HH:mm:ssZ",
-    23 -> "yyyy-MM-dd'T'HH:mm:ss.SSS",
-    28 -> "yyyy-MM-dd'T'HH:mm:ss.SSSZ"
-  ).mapValues(DateTimeFormat.forPattern)
-
-  def containsTAtFixedPosition(date: String): Boolean = {
+    24 -> "yyyy-MM-dd HH:mm:ssZ",
+    25 -> "yyyy-MM-dd HH:mm:ssZ",
+    28 -> "yyyy-MM-dd HH:mm:ss.SSSZ",
+    29 -> "yyyy-MM-dd HH:mm:ss.SSSZ"
+  )
+
+  private val formatsWithZBySize = HashMap[Int, String](
+    11 -> "yyyy-MM-ddZ", // Z is 1 length here 'Z'
+    17 -> "yyyy-MM-dd HH:mmZ",
+    20 -> "yyyy-MM-dd HH:mm:ssZ",
+    24 -> "yyyy-MM-dd HH:mm:ss.SSSZ"
+  )
+
+  /**
+    * 3D Matrix: (hasT, hasZ, length) => DateTimeFormat
+    *
+    */
+  private val converterMatrix = Array.tabulate[DateTimeFormatter](2, 2, formatsBySize.keys.max + 1) { (i, j, k) =>
+    (if (j == 0) formatsBySize else formatsWithZBySize).get(k)
+        .map(str => if (i == 1) str.replace(" ", "'T'") else str)
+        .map(DateTimeFormat.forPattern).orNull
+  }
+
+  private val slowPathConverters = (formatsBySize.values ++ formatsBySize.values.map(_.replace(" ", "'T'")))
+      .map(DateTimeFormat.forPattern)
+
+
+  private def slowParse(date: String): Date = {
+    slowPathConverters.view.map(p => Try(p.parseDateTime(date))).find(_.isSuccess) match {
+      case Some(Success(d)) => d.toDate
+      case _ => throw new IllegalArgumentException(s"Invalid date: $date")
+    }
+  }
+
+  private def containsTAtFixedPosition(date: String): Boolean = {
     date.length > 10 && date.charAt(10) == 'T'
   }
 
+  private def endsWithZ(date: String): Boolean = {
+    date.length > 0 && date.charAt(date.length - 1) == 'Z'
+  }
+
   def parse(date: String): Date = {
     try {
-      val format = if (containsTAtFixedPosition(date)) formatsWithTBySize(date.length) else formatsBySize(date.length)
+      implicit def bool2int(b: Boolean) = if (b) 1 else 0
+      // trying guess string format by length and two symbols
+      val format: DateTimeFormatter = converterMatrix(containsTAtFixedPosition(date))(endsWithZ(date))(date.length)
       format.parseDateTime(date).toDate
     } catch {
-      case _: Throwable => throw new IllegalArgumentException(s"Invalid date: [$date]")
+      // guess fail, iterate over all parsers
+      case _: Throwable => slowParse(date)
```

Comment:
If we can't find one in the conversion how will any of those match? If we couldn't tell by the hints then don't we know none of our parsers will work?

Reply:
During development I hit new TimeZone format regression two times. SO to make sure we do not create new regressions I put all old code into the slowParser. SO If I miss some exotic case, user will not get better performance but it will work as it work before (including performace)